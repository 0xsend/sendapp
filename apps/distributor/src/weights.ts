type WeightedShare = { address: `0x${string}`; amount: bigint }

export enum Mode {
  Linear = 'linear',
  Logarithmic = 'logarithmic',
  SquareRoot = 'square_root',
  Exponential = 'exponential',
  EaseInOut = 'ease_in_out',
}

export const PERC_DENOM = 1000n

export function calculatePercentageWithBips(value: bigint, bips: bigint) {
  const bps = bips * PERC_DENOM
  const percentage = value * (bps / PERC_DENOM)
  return percentage / PERC_DENOM
}

/**
 * Given a list of balances and a distribution amount, calculate the distribution weights and share amounts.
 */
export function calculateWeights(
  balances: readonly {
    address: `0x${string}`
    balance: string
  }[],
  timeAdjustedAmount: bigint,
  mode: Mode = Mode.Linear
): Record<string, WeightedShare> {
  const poolWeights: Record<`0x${string}`, bigint> = {}

  // First calculate all slashed weights
  const totalBalance = balances.reduce((acc, { balance }) => acc + BigInt(balance), 0n)

  for (const { address, balance } of balances) {
    poolWeights[address] = calculateWeightByMode(
      BigInt(balance),
      totalBalance,
      balances.length,
      mode
    )
  }

  // Calculate shares using the two sets of weights
  const weightedShares: Record<string, WeightedShare> = {}

  const totalWeight = Object.values(poolWeights).reduce((acc, w) => acc + w, 0n)

  for (const { address } of balances) {
    const poolWeight = poolWeights[address] ?? 0n

    const amount = (poolWeight * timeAdjustedAmount) / totalWeight

    if (amount > 0n) {
      weightedShares[address] = {
        amount,
        address,
      }
    }
  }

  handleRoundingErrors(weightedShares, timeAdjustedAmount)

  return weightedShares
}

// Helper function to handle rounding errors
function handleRoundingErrors(shares: Record<string, WeightedShare>, targetAmount: bigint) {
  let totalDistributed = 0n
  for (const share of Object.values(shares)) {
    totalDistributed += share.amount
  }

  if (totalDistributed > targetAmount) {
    const difference = targetAmount - totalDistributed
    const largestShare = Object.values(shares).reduce((a, b) => (a.amount > b.amount ? a : b))
    largestShare.amount += difference
  }
}

function calculateWeightByMode(
  balance: bigint,
  totalBalance: bigint,
  numBalances: number,
  mode: Mode
): bigint {
  switch (mode) {
    case Mode.Linear:
      return balance
    case Mode.Logarithmic:
      return calculateLogarithmicWeight(balance, totalBalance)
    case Mode.SquareRoot:
      return calculateSquareRootWeight(balance, totalBalance)
    case Mode.Exponential:
      return calculateExponentialWeight(balance, totalBalance, numBalances)
    case Mode.EaseInOut:
      return calculateCubicBezierWeight(balance, totalBalance)
    default:
      throw new Error(`Unknown weight mode: ${mode}`)
  }
}

function calculateLogarithmicWeight(balance: bigint, totalBalance: bigint): bigint {
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.log(1 + Number(balance) / Number(totalBalance)))
  )
}

function calculateSquareRootWeight(balance: bigint, totalBalance: bigint): bigint {
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.sqrt(1 - Number(balance) / Number(totalBalance)))
  )
}

function calculateExponentialWeight(
  balance: bigint,
  totalBalance: bigint,
  numBalances: number
): bigint {
  const k = 0.005 * numBalances
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.exp((-k * Number(balance)) / Number(totalBalance)))
  )
}

function calculateCubicBezierWeight(balance: bigint, totalBalance: bigint): bigint {
  if (totalBalance === 0n) {
    return 0n
  }

  // t_bn represents the ratio of balance/totalBalance, scaled to PERC_DENOM
  // e.g., if balance is 30% of total and PERC_DENOM is 1000000n, t_bn would be 300000n
  const t_bn = (balance * PERC_DENOM) / totalBalance

  // Cubic Bezier formula components:
  // Term 1: (1-t)³ * p0        - influence of start point
  // Term 2: 3(1-t)²t * p1      - influence of first control point
  // Term 3: 3(1-t)t² * p2      - influence of second control point
  // Term 4: t³ * p3            - influence of end point
  const p0 = 0n
  const p1 = (PERC_DENOM * 20n) / 100n // 0.2 as BigInt
  const p2 = (PERC_DENOM * 80n) / 100n // 0.8 as BigInt
  const p3 = PERC_DENOM

  // inv_t is the inverse of t_bn (distance from 1)
  // e.g., if t_bn is 300000n, inv_t would be 700000n
  const inv_t = PERC_DENOM - t_bn

  // Cubic Bezier formula components:
  // Term 1: (1-t)³ * p0        - influence of start point
  // Term 2: 3(1-t)²t * p1      - influence of first control point
  // Term 3: 3(1-t)t² * p2      - influence of second control point
  // Term 4: t³ * p3            - influence of end point
  const bezier =
    (inv_t ** 3n * p0 +
      3n * inv_t ** 2n * t_bn * p1 +
      3n * inv_t * t_bn ** 2n * p2 +
      t_bn ** 3n * p3) /
    PERC_DENOM ** 3n

  return bezier
}
