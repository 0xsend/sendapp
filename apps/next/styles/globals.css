[data-rk] {
  min-height: 100%;
}

:root {
  --sat: env(safe-area-inset-top);
  --sar: env(safe-area-inset-right);
  --sab: env(safe-area-inset-bottom);
  --sal: env(safe-area-inset-left);
}

/* Global Scrollbar Styling */
::-webkit-scrollbar {
  width: 8px;
  height: 8px;
}

::-webkit-scrollbar-thumb {
  background-color: var(--color1);
  border-radius: 4px;
}

::-webkit-scrollbar-track {
  background: transparent;
}

/* For Firefox */
body {
  scrollbar-width: thin;
  scrollbar-color: var(--color1) transparent;
}
