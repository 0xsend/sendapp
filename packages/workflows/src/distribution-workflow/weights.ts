type WeightedShare = { address: `0x${string}`; amount: bigint }
type Weights = {
  /**
   * The total weight of all balances.
   */
  totalWeight: bigint
  /**
   * The weight of each send token in the distribution.
   */
  weightPerSend: bigint
  /**
   * The weight of each balance in the distribution.
   */
  poolWeights: Record<`0x${string}`, bigint>
  /**
   * The weighted share amounts of each balance in the distribution.
   */
  weightedShares: Record<`0x${string}`, WeightedShare>
}

export enum Mode {
  Linear = 'linear',
  Logarithmic = 'logarithmic',
  SquareRoot = 'square_root',
  Exponential = 'exponential',
}

export const PERC_DENOM = 10000n

export function calculatePercentageWithBips(value: bigint, bips: bigint) {
  const bps = bips * PERC_DENOM
  const percentage = value * (bps / PERC_DENOM)
  return percentage / PERC_DENOM
}

/**
 * Given a list of balances and a distribution amount, calculate the distribution weights and share amounts.
 */
export function calculateWeights(
  balances: readonly { address: `0x${string}`; balance: string }[],
  amount: bigint,
  mode: Mode = Mode.Linear
): Weights {
  const poolWeights: Record<`0x${string}`, bigint> = {}
  const totalBalance = balances.reduce((acc, { balance }) => acc + BigInt(balance), 0n)

  for (const { address, balance } of balances) {
    const weight = calculateWeightByMode(BigInt(balance), totalBalance, balances.length, mode)
    if (poolWeights[address] === undefined) {
      poolWeights[address] = 0n
    }
    poolWeights[address] += weight
  }

  const totalWeight = Object.values(poolWeights).reduce((acc, weight) => acc + weight, 0n)
  const weightPerSend = (totalWeight * PERC_DENOM) / amount

  const weightedShares: Record<string, WeightedShare> = {}
  for (const [address, weight] of Object.entries(poolWeights)) {
    const amount = (weight * PERC_DENOM) / weightPerSend
    if (amount > 0n) {
      weightedShares[address] = {
        amount,
        address: address as `0x${string}`,
      }
    }
  }

  return { totalWeight, weightPerSend, poolWeights, weightedShares }
}

function calculateWeightByMode(
  balance: bigint,
  totalBalance: bigint,
  numBalances: number,
  mode: Mode
): bigint {
  switch (mode) {
    case Mode.Linear:
      return balance
    case Mode.Logarithmic:
      return calculateLogarithmicWeight(balance, totalBalance)
    case Mode.SquareRoot:
      return calculateSquareRootWeight(balance, totalBalance)
    case Mode.Exponential:
      return calculateExponentialWeight(balance, totalBalance, numBalances)
    default:
      throw new Error(`Unknown weight mode: ${mode}`)
  }
}

function calculateLogarithmicWeight(balance: bigint, totalBalance: bigint): bigint {
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.log(1 + Number(balance) / Number(totalBalance)))
  )
}

function calculateSquareRootWeight(balance: bigint, totalBalance: bigint): bigint {
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.sqrt(1 - Number(balance) / Number(totalBalance)))
  )
}

function calculateExponentialWeight(
  balance: bigint,
  totalBalance: bigint,
  numBalances: number
): bigint {
  const k = 0.005 * numBalances
  return BigInt(
    Math.floor(Number(PERC_DENOM) * Math.exp((-k * Number(balance)) / Number(totalBalance)))
  )
}
