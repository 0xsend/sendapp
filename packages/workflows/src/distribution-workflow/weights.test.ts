import { describe, expect, it } from '@jest/globals'

import { calculateWeights, type Mode } from './weights'
const balances = [
  { address: '0x1', balance: '1000' },
  { address: '0x2', balance: '2000' },
  { address: '0x3', balance: '3000' },
] as const
const amount = 100000n

const testCases = [
  {
    name: 'linear',
    mode: 'linear',
    expected: {
      totalWeight: 6000n,
      weightPerSend: 600n,
      poolWeights: {
        '0x1': 1000n,
        '0x2': 2000n,
        '0x3': 3000n,
      },
      weightedShares: {
        '0x1': {
          amount: 16666n,
          address: '0x1',
        },
        '0x2': {
          amount: 33333n,
          address: '0x2',
        },
        '0x3': {
          amount: 50000n,
          address: '0x3',
        },
      },
    },
  },
  {
    name: 'logarithmic',
    mode: 'logarithmic',
    expected: {
      totalWeight: 8471n,
      weightPerSend: 847n,
      poolWeights: {
        '0x1': 1541n,
        '0x2': 2876n,
        '0x3': 4054n,
      },
      weightedShares: {
        '0x1': {
          address: '0x1',
          amount: 18193n,
        },
        '0x2': {
          address: '0x2',
          amount: 33955n,
        },
        '0x3': {
          address: '0x3',
          amount: 47863n,
        },
      },
    },
  },
  {
    name: 'square_root',
    mode: 'square_root',
    expected: {
      totalWeight: 24363n,
      weightPerSend: 2436n,
      poolWeights: {
        '0x1': 9128n,
        '0x2': 8164n,
        '0x3': 7071n,
      },
      weightedShares: {
        '0x1': {
          address: '0x1',
          amount: 37471n,
        },
        '0x2': {
          address: '0x2',
          amount: 33513n,
        },
        '0x3': {
          address: '0x3',
          amount: 29027n,
        },
      },
    },
  },
  {
    name: 'exponential',
    mode: 'exponential',
    expected: {
      totalWeight: 29850n,
      weightPerSend: 2985n,
      poolWeights: {
        '0x1': 9975n,
        '0x2': 9950n,
        '0x3': 9925n,
      },
      weightedShares: {
        '0x1': {
          address: '0x1',
          amount: 33417n,
        },
        '0x2': {
          address: '0x2',
          amount: 33333n,
        },
        '0x3': {
          address: '0x3',
          amount: 33249n,
        },
      },
    },
  },
] as {
  name: string
  mode: Mode
  expected: {
    totalWeight: bigint
    weightPerSend: bigint
    poolWeights: Record<`0x${string}`, bigint>
    weightedShares: Record<`0x${string}`, { address: `0x${string}`; amount: bigint }>
  }
}[]

describe('calculateWeights', () => {
  for (const { name, mode, expected } of testCases) {
    it(`should calculate ${name} weights using ${mode} mode`, () => {
      const snapshot = calculateWeights(balances, amount, mode)

      expect(snapshot).toMatchSnapshot(mode)

      const { totalWeight, weightPerSend, poolWeights, weightedShares } = snapshot

      expect(totalWeight).toBe(expected.totalWeight)
      expect(weightPerSend).toBe(expected.weightPerSend)
      expect(poolWeights).toEqual(expected.poolWeights)
      expect(weightedShares).toEqual(expected.weightedShares)
    })
  }
})
