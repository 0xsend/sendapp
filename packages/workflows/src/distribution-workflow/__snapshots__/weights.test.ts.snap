// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateWeights should calculate exponential weights using exponential mode: exponential 1`] = `
{
  "poolWeights": {
    "0x1": 9975n,
    "0x2": 9950n,
    "0x3": 9925n,
  },
  "totalWeight": 29850n,
  "weightPerSend": 2985n,
  "weightedShares": {
    "0x1": {
      "address": "0x1",
      "amount": 33417n,
    },
    "0x2": {
      "address": "0x2",
      "amount": 33333n,
    },
    "0x3": {
      "address": "0x3",
      "amount": 33249n,
    },
  },
}
`;

exports[`calculateWeights should calculate linear weights using linear mode: linear 1`] = `
{
  "poolWeights": {
    "0x1": 1000n,
    "0x2": 2000n,
    "0x3": 3000n,
  },
  "totalWeight": 6000n,
  "weightPerSend": 600n,
  "weightedShares": {
    "0x1": {
      "address": "0x1",
      "amount": 16666n,
    },
    "0x2": {
      "address": "0x2",
      "amount": 33333n,
    },
    "0x3": {
      "address": "0x3",
      "amount": 50000n,
    },
  },
}
`;

exports[`calculateWeights should calculate logarithmic weights using logarithmic mode: logarithmic 1`] = `
{
  "poolWeights": {
    "0x1": 1541n,
    "0x2": 2876n,
    "0x3": 4054n,
  },
  "totalWeight": 8471n,
  "weightPerSend": 847n,
  "weightedShares": {
    "0x1": {
      "address": "0x1",
      "amount": 18193n,
    },
    "0x2": {
      "address": "0x2",
      "amount": 33955n,
    },
    "0x3": {
      "address": "0x3",
      "amount": 47863n,
    },
  },
}
`;

exports[`calculateWeights should calculate square_root weights using square_root mode: square_root 1`] = `
{
  "poolWeights": {
    "0x1": 9128n,
    "0x2": 8164n,
    "0x3": 7071n,
  },
  "totalWeight": 24363n,
  "weightPerSend": 2436n,
  "weightedShares": {
    "0x1": {
      "address": "0x1",
      "amount": 37471n,
    },
    "0x2": {
      "address": "0x2",
      "amount": 33513n,
    },
    "0x3": {
      "address": "0x3",
      "amount": 29027n,
    },
  },
}
`;
